cep-recipe-jinja.md

```diff
@@ -0,0 +1,320 @@
+
+A Jinja expression in the new recipe format looks like the following:
+
+`${{ version }}`
+
+Or if a function is involved:
+
+`${{ compiler('c') }}`
+
+Jinja expressions are also used in `if` statements and in the `skip` field of a recipe. In both instances, the `${{ ... }}` syntax is omitted.
+
+E.g.:
+
+```yaml
+build:
+  skip:
+    - osx  # This is a Jinja expression!
+
+requirements:
+  build:
+    - if: win and cuda  # This is a Jinja expression!
+      then:
+        - cudatoolkit
+```
+
+## Variables in the recipe
+
+The variables that are available in the Jinja context in the recipe come from two sources: the "variant configuration" file or the "context" section of the recipe.
+
+### The context section
+
+The context is a dictionary at the top-level of a recipe that maps keys to scalar values. The keys can be used by accessing them as variables in the Jinja expressions. For example:
+
+```yaml
+context:
+  version: "1.0.5"
+
+package:
+  version: ${{ version }}
+```
+
+Context evaluation happens from top-to-bottom. That means a later value can reference an earlier one like so:
+
+```yaml
+context:
+  version: "1.0.5"
+  name_and_version: "pkg_${{ version }}"
+```
+
+### Variables from variant configuration
+
+Any variable specified in the variant configuration file can be used in a recipe by using it in a Jinja expression. For example, with a variant config like:
+
+```yaml
+cuda:
+  - yes
+  - no
+```
+
+This value can be used as follows, for example in an inline `if` expression:
+
+```yaml
+requirements:
+  host:
+    - ${{ "cudatoolkit" if cuda == "yes" }}
+```
+
+## Available Jinja functions
+
+## The compiler function
+
+The compiler function is used to stick together a compiler from `{lang}_compiler` and `{lang}_compiler_version`
+
+The function looks as follows:
+
+```yaml
+${{ compiler('c') }}
+```
+
+This would pull in the c_compiler and c_compiler_version from the variant config. The compiler function suffixes `{lang}_compiler` with the `target_platform` to render to something such as:
+
+```
+gcc_linux-64 8.9
+clang_osx-arm64 12
+msvc_win-64 19.29
+```
+
+The function `${{ compiler("foo") }}` thus evaluates to `{foo_compiler}_{target_platform} {foo_compiler_version}`.
+
+The variant config could look something like:
+
+```yaml
+c_compiler:
+  - gcc
+c_compiler_version:
+  - "8.9"
+cxx_compiler:
+  - clang
+cxx_compiler_version:
+  - "12"
+```
+
+> [!NOTE]
+> Default values for the `compiler` function
+>
+> If not further specified, the following values are used as default values:4
+>
+> ```yaml
+> linux:
+>  c: gcc
+>  cxx: gxx
+>  fortran: gfortran
+>  rust: rust
+> osx:
+>   c: clang
+>   cxx: clangxx
+>   fortran: gfortran
+>   rust: rust
+> win:
+>   c: vs2017
+>   cxx: vs2017
+>   fortran: gfortran
+>   rust: rust
+> ```
+
+## The `stdlib` function
+
+The `stdlib` function works exactly as the `compiler` function.
```

Comment:
Not sure if that's the intention here, but perhaps a bit of clarification nevertheless.
```suggestion
The `stdlib` function works exactly as the `compiler` function, except that it has not pre-defined defaults.
```

Reply:
Yeah this section is not complete yet. I think we should define defaults for stdlib? We need to make the experience of conda recipes better...

Reply:
> Yeah this section is not complete yet. 

Ah, that wasn't clear to me. I'm ambivalent about the usefulness of defaults, but if your goal is that people are able to write recipes outside of cf (where we have the defaults in the global pinning anyway) & Anaconda main channels _and_ not have to worry about where their compiler is coming from, then I guess that's a usecase.

Reply:
>I think I would like to have an awesome "default" experience. If you take a recipe from conda-forge, ideally you can just run build on it without having to worry about the variant config file.

The thing is that this is very ecosystem specific. The default might work for some time if conda-forge uses the same defaults, but as soon as an ecosystem like conda-forge diverges from the CEP defaults, then the recipes might not be buildable without a variant config. Same for the defaults channel.

Reply:
I am wondering some more abotu the "no defaults". E.g. if I take a feedstock, and want to add support for macOS - it's pretty annoying because I don't have a .ci_config for osx-arm64 yet. It would be most convenient to either use a default. It would be more correct to point to the `conda-forge-pinning-feedstock`. 

I think ideally we publish the pinnings next to the repodata file or something like that.